### DISCUSSION TITLE
Adding enums in schema – ConversionError(Enum value not stored as enum)

### DISCUSSION THREAD

**Discussion by bexuma** [prisma/prisma#4149] [Q&A] [answered]
### Background
I am migrating my GraphQL server from Ruby on Rails to Node.js with the help of Apollo Server and Prisma. It utilizes PostgreSQL database.

Today is my 2nd day of exploring Prisma. I have already introspected the database, run some queries against it and renamed the schema models to match singular PascalCase naming convention.

### Status Quo
Below you could follow the current flow I use for Enums.

#### schema.prisma

```javascript
model User {
  id        Int      @id @default(autoincrement())
  status    Int?     @default(0)
}
```

#### script.js

```javascript
const typeDefs = gql`
  type Query {
    firstUser: User
  }

  type User {
    id: ID
    status: UserStatus
  }

  enum UserStatus {
    student
    generalSpecialist
    dormSpecialist
    dormSupervisor
  }
`;

const resolvers = {
  Query: {
    firstUser: async () => await prisma.user.findFirst();
  },
  UserStatus: {
    student: 0,
    generalSpecialist: 1,
    dormSpecialist: 2,
    dormSupervisor: 3,
  }
};
```

### Room for improvement
It works just fine, but I want to add enums in schema as well. However, I get _**ConversionError(Enum value not stored as enum)**_ when I try doing so (the code snippet is below). What am I doing wrong?

**schema.prisma**

```javascript
model User {
  id        Int           @id @default(autoincrement())
  status    UserStatus?   @default(student)
}

enum UserStatus {
  student
  generalSpecialist
  dormSpecialist
  dormSupervisor
}
```

#### Update after @themmyloluwaa's answer
I tried capitalizing the enum's values, getting same error.

**Comment by themmyloluwaa** [comment]
You can definitely add enum values in schema with the enum keyword. Taking your example, below your User model

```javascript
model User {
  id        Int      @id @default(autoincrement())
  status    Int?     @default(0)
}
```
add your enums as

```javascript
enum UserStatus: {
    STUDENT,
    GENERALSPECIALIST,
    DORMSPECIALIST,
    DORMSUPERVISOR,
  }
```
Then reference the default on your user model as

```javascript
model User {
  id        Int      @id @default(autoincrement())
  status    UserStatus?     @default(STUDENT)
}
```
Try this out

**Reply by bexuma** [reply]
You have pasted the same code I included in **Room for improvement**, but with enum values being capitalized. I know, it's the convention, but for now I stick to lowercase because of the frontend I have which assumes the enums are lowercased.

Nevertheless I tried making the enum's values all capital after your suggestion, but it is still giving me the same error.

**Reply by ryands17** [reply]
@bexuma When you introspected the DB, were the enums also written to your `schema.prisma`? If you could provide the SQL of your table and enums I can try reproducing this.

**Reply by bexuma** [reply]
Only types were generated on introspection. **student** column is basically of an integer type. I figured, that is exactly what causes the issue.

#### "The problem is with Rails"
Instead of creating enumerated types on migrations, Rails encourages the use of integer based enums. In Rails, you handle enums by providing a map linking keys with integers.

### What's next?
Does Prisma provide a level of abstraction as Rails does by configuring the map linking integers with keys? It would be troublesome to filter users by status if I need to pass in integer values in Prisma Client.

#### Alternative
I can try changing data type of **status** from integer to enum in the database, though it does not seem to be an optimal solution.

**Reply by ryands17** [reply]
@bexuma Currently this abstraction is not available to map integers directly with keys. You would need to change the types for now as a workaround.

I would suggest opening a feature request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) for the above so that we can consider adding a polyfill for this type of conversion.

**Reply by ryands17** [reply, marked as answer]
@bexuma Added a feature request for the same https://github.com/prisma/prisma/issues/4169

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
